docs($compile): refine explanation of `=`-bindings

thorn0 opened the pull request:
The current version of this paragraph is in many ways inaccurate and confusing.

gkalpak commented on src/ng/compile.js:
```diff
@@ -174,17 +174,18 @@
- *   `=*` or `=*attr` (`=*?` or `=*?attr` if the property is optional).
+ * * `=` or `=attr` - set up a bidirectional binding between a local scope property and an expression
+ *   passed via the attribute `attr`. The expression is evaluated in the context of the parent scope and
+ *   can be either an assignable expression or a literal value. If no `attr` name is specified then the
```
Since you are at it, it might be a good idea so somehow mention that if a non-assignable expression is passed, then `$compile:nonassign` will be thrown when trying to assign to it.

gkalpak commented on src/ng/compile.js:
```diff
@@ -174,17 +174,18 @@
+ *   reflected in `localModel` and vice versa. If the attribute doesn't exist, an exception will be thrown
+ *   ({@link error/$compile/nonassign $compile:nonassign}). You can avoid this behavior and flag the
+ *   property as optional using `=?` or `=?attr`. By default, the {@link ng.$rootScope.Scope#$watch} method
+ *   is used, the equality check is done by reference. If you need to watch a collection for changes using
```
_is used, the equality_  -->  _is used and the equality_ (?)

gkalpak replied:
"watch a collection" is a little vague (users might wonder if a collection is something angular specific). I slightly preferred the previous phrasing.

thorn0 replied:
The previous phrasing is really misleading. "If you want to _shallow_ watch for changes" sounds like by default the watch is deep whereas it's not the case.

gkalpak replied:
Fair enough :smiley:

thorn0 commented:
Is it looking better now?

gkalpak commented on src/ng/compile.js:
```diff
@@ -174,17 +174,20 @@
+ *   attribute name is assumed to be the same as the local name. Given `<widget my-attr="parentModel">`
+ *   and the widget definition `scope: { localModel:'=myAttr' }`, the property `localModel` on the widget's
+ *   scope will reflect the value of `parentModel` on the parent scope. Changes to `parentModel` will be
+ *   reflected in `localModel` and vice versa. If the attribute doesn't exist, an exception will be thrown
```
Actually, I think it will only throw if you try to **assign** and the expression is non-assignable or the attribute is missing.

thorn0 replied:
Right. I believe that's exactly what I wrote.

gkalpak replied:
I'm not sure I was clear. I meant it will throw if:

1. The expression is non-assignable and you try to assign to it.
2. The attribut is missing **and you try to assign to it**.

I believe that the part in bold is missing from the description.

thorn0 replied:
Ah, I see. You're right. But is it worth going into these details when all we need here is just to give a notion of optional bindings? Won't it overcomplicate the explanation?

However, I found another interesting detail missing from this text. Namely, when the expression is an object or array literal, the watch is deep.

thorn0 replied:
I mean, if the attribute is optional, please mark it as such. The idea should be like this. It doesn't really matter how and when exactly it fails if you didn't mark it.

thorn0 commented:
@gkalpak pls have a look

gkalpak commented on src/ng/compile.js:
```diff
@@ -174,17 +174,22 @@
- *   `=*` or `=*attr` (`=*?` or `=*?attr` if the property is optional).
+ * * `=` or `=attr` - set up a bidirectional binding between a local scope property and an expression
+ *   passed via the attribute `attr`. The expression is evaluated in the context of the parent scope and
+ *   must be either an assignable expression or a literal value. If no `attr` name is specified then the
```
It is not required to be an assignable expression. Only if you try to assign to it :)

thorn0 replied:
Done

Narretz commented:
Looks good. One suggestion, can you change all instances of widget to directive? I widget is a relict of very old times, iirc, back before the term directive was introduced.

Narretz commented on src/ng/compile.js:
```diff
@@ -174,17 +174,22 @@
+ *   doesn't exist, an exception ({@link error/$compile/nonassign `$compile:nonassign`}) will be
+ *   thrown on discovering changes to the local value since it'll be impossible to sync it with the
+ *   parent scope in this case. By default, the {@link ng.$rootScope.Scope#$watch `$watch`} method is
+ *   used for tracking changes, and the equality check is done by reference. However, if an object
```
Personally, I find it clearer to say that the equality check is based on object identity. I guess it means the same though.

thorn0 commented:
> can you change all instances of widget to directive?

What about "component"? It's already called so a bit above:

> **`{...}` (an object hash):** A new "isolate" scope is created for the directive's element. [...] This is useful when creating reusable components...

Narretz commented:
I would be okay with `<my-component my-attr="parentModel">`, but `directive definition` and `directive's scope`. Component is a rather specific term. I think the part you quoted shows that. Components imply a certain modularity, but directives can also add events etc. without being components.

thorn0 committed:
docs($compile): refine explanation of `=`-bindings

The current version of this paragraph is in many ways inaccurate and confusing.

thorn0 commented:
OK. Eliminated widgets and replaced 'by reference' with 'based on object identity'.

But I also don't like that it says "local scope", "directive's scope" everywhere whereas the same binding mechanism is used for controllers.

Narretz commented:
I think that's okay. The text is very information dense anyway. bindToController should make clear that it changes this behavior.

thorn0 commented:
I think bindings should get a higher level section of their own, like [Transcusion](https://docs.angularjs.org/api/ng/service/$compile#transclusion). It can be done later though.

gkalpak commented:
LGTM

Narretz commented:
@gkalpak Can you merge? Then I can rebase the one-way PR.

gkalpak commented:
@Narretz: Merged.
(Backported to `v1.4.x` as 8a1f600.)
